cpp/src/arrow/gpu/cuda_context.h

```diff
@@ -119,6 +119,12 @@ class ARROW_EXPORT CudaContext : public std::enable_shared_from_this<CudaContext
   /// \brief Return device number
   int device_number() const;
 
+  /// \brief Get device address for the context
+  /// \param[in] addr device address
+  /// \param[out] devaddr the device address for the context
+  /// \return Status
+  Status GetDeviceAddress(uintptr_t addr, uintptr_t* devaddr);
```

Comment:
The first argument must be `uintptr_t` as it holds memory address value. The first and second arguments are closely related (in most cases `addr == *devaddr`). So, using `uintptr_t*` for the second argument for consistency.

Another approach would be to declare `Status GetDeviceAddress(const void* addr, void** devaddr)` but how this would be better in terms of clarity or safety or performance?

Reply:
> The first argument must be uintptr_t as it holds memory address value.

A memory address value is a pointer value. So `void*` (not `const void*`) would work as well. With `uintptr_t` you're bound to do some casting manually when calling this method.

(actually, in Arrow we generally use `uint8_t`*)

> how this would be better in terms of clarity or safety or performance?

I think regular pointers would be clearer. Performance would obviously be the same.

Reply:
I agree that "a memory address value is a pointer value" but this relation is commutative for host memory and for device memory addresses the statement becomes invalid in CPU functions.

Holding a device memory address as a host pointer value sounds like an abuse of pointer notion as the pointer value would point to invalid location in host memory. I know this is how arrow `CudaBuffer` holds the device memory address value but that's clear when using it within `CudaBuffer`. In other places (that is not device kernel, for instance), dereferencing such a pointer value will lead to segfaults.

Ideally, a library that deals with devices that have separate memory address spaces from the host memory, should make really hard to crash user programs. Making a distinction of memory address value and the pointer value would contribute to this ideal, IMHO.
I have been experimenting with this idea within [umem](https://github.com/plures/umem) but the connection to arrow might require a broader discussion than this PR. CC: @wesm

Reply:
You have a point :-) However, until now our convention is to use `void*` or `uint8_t*` for pointers to arbitrary data. Even `CudaBuffer::CopyFromDevice` takes a `void*` for the source device address. So it seems like we're introducing an inconsistency, while not really protecting users.

Trying to bring more safety in buffer typing should be part of the [ARROW-2447](https://issues.apache.org/jira/browse/ARROW-2447) design/discussion, I think.

Reply:
OK, will use `void*` for consistency and continue discussion on memory address representation in ARROW 2447.